### DISCUSSION TITLE
signInWithMicrosoft: Web layer authentication

### DISCUSSION THREAD

**Discussion by bastifix** [capawesome-team/capacitor-firebase#165] [Q&A]
Hi, I want to use Microsoft Azure AD Accounts to login into the web layer. For Google-Login best practice seems to use code like this: 

```
const signInWithGoogle = async () => {
  // 1. Create credentials on the native layer
  const result = await FirebaseAuthentication.signInWithGoogle();
  // 2. Sign in on the web layer using the id token
  const credential = GoogleAuthProvider.credential(result.credential?.idToken);
  const auth = getAuth();
  auth.onAuthStateChanged(user => {
    console.log('User: ' + user);
  });
  await signInWithCredential(auth, credential);
};
```

Unfortunately Microsoft doesn't support `signInWIthCredential` [https://cloud.google.com/identity-platform/docs/web/microsoft?hl=en](https://cloud.google.com/identity-platform/docs/web/microsoft?hl=en) --> so it seems not possible for me to sign in to the web layer?

Here: [https://stackoverflow.com/questions/63678222/how-to-implement-microsoft-login-with-firebase-in-react-native#63693445](https://stackoverflow.com/questions/63678222/how-to-implement-microsoft-login-with-firebase-in-react-native#63693445)
is suggested to use `signInWithCustomToken` but it was not possible for me to create a customToken while using a callable cloud function from the not signed in user. Here my cloud function code:

```
exports.getCustomToken = functions.https.onCall((data, context) => {
    return new Promise((resolve, reject) => {
        admin.auth().createCustomToken(data.uid)
            .then(response => {
                console.log('Response', response);
                resolve(response);
            })
            .catch(error => {
                console.log('Error', error);
                reject(error);
            })
    });
});
```

Calling this function ended in an error 302.

Would love some ideas how to sign users in to the web layer with Microsoft ;-)

**Comment by robingenz** [comment]
Hi @bastifix,
Have you tried this?

```ts
const signInWithMicrosoft = async () => {
  // 1. Create credentials on the native layer
  const result = await FirebaseAuthentication.signInWithMicrosoft();
  // 2. Sign in on the web layer using the id token and nonce
  const provider = new OAuthProvider('microsoft.com');
  const credential = provider.credential({
    idToken: result.credential?.idToken,
  });
  const auth = getAuth();
  auth.onAuthStateChanged(user => {
    console.log('User: ' + user);
  });
  await signInWithCredential(auth, credential);
};
```

**Reply by bastifix** [reply]
Yes, but I get in the console:
`Failed to load resource: the server responded with a status of 400 ()`

And the exact error is: 
```
{
    "error": {
        "code": 400,
        "message": "INVALID_CREDENTIAL_OR_PROVIDER_ID : Invalid IdP response/credential: http://localhost?&id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjJaUXBKM1VwYmpBWVhZR2FYRUpsOGxWMFRPSSJ9.eyJhdWQiOiJjZDYwZGIwYy1hYzYzLTQyYWMtYmZmOS05YjJlMDdiY2RiNWIiLCJpc3MiOiJodHRwczovL2xvZ2luLm1pY3Jvc29mdG9ubGluZS5jb20vZjRjYzg2YWMtYzZjNS00MjUyLWI0ZTgtZWIyYjgwODYzMzRiL3YyLjAiLCJpYXQiOjE2NTkzNjI2MjksIm5iZiI6MTY1OTM2MjYyOSwiZXhwIjoxNjU5MzY2NTI5LCJuYW1lIjoiVGVzdGVyIiwib2lkIjoiYTkxYmNkYjQtZTFiOC00MmNjLWFiYWUtZmQxNjQ5ZDJiYzk2IiwicHJlZmVycmVkX3VzZXJuYW1lIjoidGVzdGVyQHBhcmFnb24tZGlzLmNvbSIsInJoIjoiMC5BVTRBckliTTlNWEdVa0swNk9zcmdJWXpTd3piWU0xanJLeEN2X21iTGdlODIxdE9BSjguIiwic3ViIjoiOURmX3UwdzIwb0l2QlJTcDZzTTJTeWVBbWhUVFh4a0VLeVh0OU9zTThqNCIsInRpZCI6ImY0Y2M4NmFjLWM2YzUtNDI1Mi1iNGU4LWViMmI4MDg2MzM0YiIsInV0aSI6Ikc0eHVkMUFVTlV5UXNpSzVpbTlHQVEiLCJ2ZXIiOiIyLjAifQ.rPmOPRCv90bkXuFDExdgREt9C5dIVyXaW6IyiugKyVKMV6FGX51JXEQ5hRrpkgSeuzv8uJwTMN6y0IGCwob7XmgDj91XHRHsLwtcnhGvHyATLuLXLv_VGCtq2Gb-84sEMT2OWWdmFG_h8bJVuDbYny5Ypz-ZsiaYQkW-ohC7vTA01OGuhcqz2C9vXkXxSYH7L9Khn8tDGolkJOU0c5pZLmbuvicEfyaQ4Iwxk6YZvq3f7Jeff0Tg7yKGlsw_mRQbVxEPkEcvPpcDsjW-73XQhOKOQ4lgN8up2XmfUiQJv6la9aFUngWIwGkTD6e0UbVeTbyKswpWi3LJmP-fuJSyzA&providerId=microsoft.com",
        "errors": [
            {
                "message": "INVALID_CREDENTIAL_OR_PROVIDER_ID : Invalid IdP response/credential: http://localhost?&id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjJaUXBKM1VwYmpBWVhZR2FYRUpsOGxWMFRPSSJ9.eyJhdWQiOiJjZDYwZGIwYy1hYzYzLTQyYWMtYmZmOS05YjJlMDdiY2RiNWIiLCJpc3MiOiJodHRwczovL2xvZ2luLm1pY3Jvc29mdG9ubGluZS5jb20vZjRjYzg2YWMtYzZjNS00MjUyLWI0ZTgtZWIyYjgwODYzMzRiL3YyLjAiLCJpYXQiOjE2NTkzNjI2MjksIm5iZiI6MTY1OTM2MjYyOSwiZXhwIjoxNjU5MzY2NTI5LCJuYW1lIjoiVGVzdGVyIiwib2lkIjoiYTkxYmNkYjQtZTFiOC00MmNjLWFiYWUtZmQxNjQ5ZDJiYzk2IiwicHJlZmVycmVkX3VzZXJuYW1lIjoidGVzdGVyQHBhcmFnb24tZGlzLmNvbSIsInJoIjoiMC5BVTRBckliTTlNWEdVa0swNk9zcmdJWXpTd3piWU0xanJLeEN2X21iTGdlODIxdE9BSjguIiwic3ViIjoiOURmX3UwdzIwb0l2QlJTcDZzTTJTeWVBbWhUVFh4a0VLeVh0OU9zTThqNCIsInRpZCI6ImY0Y2M4NmFjLWM2YzUtNDI1Mi1iNGU4LWViMmI4MDg2MzM0YiIsInV0aSI6Ikc0eHVkMUFVTlV5UXNpSzVpbTlHQVEiLCJ2ZXIiOiIyLjAifQ.rPmOPRCv90bkXuFDExdgREt9C5dIVyXaW6IyiugKyVKMV6FGX51JXEQ5hRrpkgSeuzv8uJwTMN6y0IGCwob7XmgDj91XHRHsLwtcnhGvHyATLuLXLv_VGCtq2Gb-84sEMT2OWWdmFG_h8bJVuDbYny5Ypz-ZsiaYQkW-ohC7vTA01OGuhcqz2C9vXkXxSYH7L9Khn8tDGolkJOU0c5pZLmbuvicEfyaQ4Iwxk6YZvq3f7Jeff0Tg7yKGlsw_mRQbVxEPkEcvPpcDsjW-73XQhOKOQ4lgN8up2XmfUiQJv6la9aFUngWIwGkTD6e0UbVeTbyKswpWi3LJmP-fuJSyzA&providerId=microsoft.com",
                "domain": "global",
                "reason": "invalid"
            }
        ]
    }
}
```

**Reply by robingenz** [reply]
Have you also tried it with `accessToken` or `idToken` and `accessToken`?

```ts
const signInWithMicrosoft = async () => {
  // 1. Create credentials on the native layer
  const result = await FirebaseAuthentication.signInWithMicrosoft();
  // 2. Sign in on the web layer using the access token
  const provider = new OAuthProvider('microsoft.com');
  const credential = provider.credential({
    accessToken: result.credential?.accessToken,
  });
  const auth = getAuth();
  auth.onAuthStateChanged(user => {
    console.log('User: ' + user);
  });
  await signInWithCredential(auth, credential);
};
```

Maybe in combination with the [`skipNativeAuth`](https://github.com/capawesome-team/capacitor-firebase/tree/main/packages/authentication#configuration) configuration option?

**Reply by bastifix** [reply]
Yes, the error is still the same...

**Reply by bastifix** [reply]
Do you think there's another way?
Or the workaround with the customToken should work somehow?
Thanks!

**Reply by robingenz** [reply]
I have to debug it. It is still on my todo list. I am a bit busy right now with the Capacitor 4 updates.

**Reply by bastifix** [reply]
I found the same error [here](https://groups.google.com/g/firebase-talk/c/Uai7n_Z-w_Y), but I'm not sure if the solution is leading somewhere:
```
The problem was I did not include the tenant id in the token/authentication endpoint. In the docs [1]

They put the token endpoint as: https://login.microsoftonline.com/common/oauth2/v2.0/authorize

but really it should be [https://login.microsoftonline.com/{tendantId}](https://login.microsoftonline.com/%7BtendantId%7D) /oauth2/v2.0/authorize

Same goes for the authorization endpoint should be:
[https://login.microsoftonline.com/{tenantId}/oauth2/v2.0/authorize](https://login.microsoftonline.com/%7BtenantId%7D/oauth2/v2.0/authorize)"

I don't understand why some of their docs have common, and others have tenantId...or why theres's a difference. 

[1] https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow
```

**Reply by robingenz** [reply]
Hi @bastifix,
first of all: Thank you for your sponsorship! That means a lot to us!
I wish I had better news but unfortunately it seems that your assumption is correct.
As mentioned [here](https://cloud.google.com/identity-platform/docs/web/microsoft?hl=en), you can indeed not use the Microsoft sign-in with `signInWithCredential`.

>You can then use [Custom authentication](https://cloud.google.com/identity-platform/docs/web/custom) to exchange the Microsoft credential for a custom token.6

Unfortunately, I have no experience with this yet.

I will try to transfer [this guide](https://firebase.google.com/docs/auth/web/cordova) to Capacitor in the next weeks. Maybe this will make the whole process with the Firebase JS SDK easier and you can use Microsoft. But before that I have to finish two other projects and update all plugins to Capacitor 4.

**Reply by bastifix** [reply]
Thanks! I'm curious to see if it will work soon ;-)

**Reply by colinethrl** [reply]
Hello! I was redirected here as [my issue](#425) was a duplicate (sorry for that). Is there any news on this?
I tried using [Custom authentication](https://cloud.google.com/identity-platform/docs/web/custom) but `getAuth().createCustomToken(uid)` doesn't exist as getAuth() from "firebase/auth" is of type Auth.
I also tried following [this guide](https://firebase.google.com/docs/auth/web/cordova) but as I'm on Capacitor I cannot install all the necessary plugins (some of them can be installed with npm but for example browsertab cannot). So I always end up with this error:
`firebaseerror: firebase: the following cordova plugins must be installed to enable oauth sign-in: cordova-plugin-buildinfo, cordova-universal-links-plugin, cordova-plugin-browsertab, cordova-plugin-inappbrowser and cordova-plugin-customurlscheme`
FYI, I'm on iOS, ionic 7.1.1 and capacitor 5.
Thanks a lot for the time you already spent on this, I hope a solution will be found!

**Reply by robingenz** [reply]
As mentioned in #425, I am currently working on supporting all Firebase SDKs so that the Firebase JS SDK on Android and iOS is no longer necessary.

**Reply by colinethrl** [reply]
All right, thanks a lot for you work! Keep in touch then!

**Reply by DavidVorac** [reply]
Hi @robingenz, first of all thanks for your hard work!

I am currently also trying to add Microsoft provider, but that would mean creating a custom token which is a lengthy process. I wonder if you have some approximate release date on supporting all Firebase SDKs?

Thanks in advance and good luck!

**Reply by robingenz** [reply]
@DavidVorac No, unfortunately there is no ETA. Cloud Firestore and Cloud Storage should be released this year.

**Reply by Ludovic1974** [reply]
Hi @robingenz 
I'm trying to implement init session using the signInMicrosoft() method but I have the following error: 
"An error occurred while trying to obtain the hash of your package's certificate". 
Is there some fix about this issue?

**Reply by robingenz** [reply]
@Ludovic1974 This error is related to your configuration and comes from the Firebase SDK. This cannot be fixed in the plugin.
